core/src/main/java/discord4j/core/object/Template.java

```diff
@@ -0,0 +1,214 @@
+package discord4j.core.object;
+
+import discord4j.common.util.Snowflake;
+import discord4j.core.GatewayDiscordClient;
+import discord4j.core.object.entity.Guild;
+import discord4j.core.object.entity.User;
+import discord4j.core.spec.GuildTemplateEditSpec;
+import discord4j.core.spec.TemplateCreateGuildSpec;
+import discord4j.discordjson.json.SerializedSourceGuildData;
+import discord4j.discordjson.json.TemplateData;
+import reactor.core.publisher.Mono;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public class Template implements DiscordObject {
+
+    /** The gateway associated to this object. */
+    private final GatewayDiscordClient gateway;
+
+    /** The raw data as represented by Discord. */
+    private final TemplateData data;
+
+    /** The ID of the guild this template is associated to. */
+    private final long guildId;
+
+    /**
+     * Constructs a {@code Template} with an associated Discord client and data.
+     *
+     * @param gateway The {@link GatewayDiscordClient} associated to this object, must be non-null.
+     * @param data The raw data as represented by Discord, must be non-null.
+     */
+    public Template(final GatewayDiscordClient gateway, final TemplateData data) {
+        this.gateway = Objects.requireNonNull(gateway);
+        this.data = Objects.requireNonNull(data);
+        this.guildId = Snowflake.asLong(data.sourceGuildId());
+    }
+
+    @Override
+    public GatewayDiscordClient getClient() {
+        return this.gateway;
+    }
+
+    /**
+     * Returns this object underlying raw data structure.
+     *
+     * @return an immutable data representation of this object
+     */
+    public TemplateData getData() {
+        return data;
+    }
+
+    /**
+     * Returns the template code (unique ID).
+     *
+     * @return the template code (unique ID)
+     */
+    public final String getCode() {
```

Comment:
I don't think these methods need to be final. Most of our entity classes are final, though.